gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/SystemTaskUpgrader.java

```diff
@@ -0,0 +1,132 @@
+
+    @Lazy
+    @Autowired
+    protected SystemTaskRepository systemTaskRepository;
+
+    @Override
+    public boolean upgrade() {
+        final String instanceOperationId = UUID.randomUUID().toString();
+        final String taskId = getTaskId();
+        boolean upgraded = systemTaskRepository.findById(taskId)
+                .switchIfEmpty(Single.defer(() -> createSystemTask(instanceOperationId)))
+                .flatMap(task -> {
+                    switch (SystemTaskStatus.valueOf(task.getStatus())) {
+                        case INITIALIZED:
+                            return processUpgrade(instanceOperationId, task, instanceOperationId);
+                        case FAILURE:
+                            // In Failure case, we use the operationId of the read task otherwise update will always fail
+                            // force the task.operationId to assign the task to the instance
+                            String previousOperationId = task.getOperationId();
+                            task.setOperationId(instanceOperationId);
+                            return processUpgrade(instanceOperationId, task, previousOperationId);
+                        case ONGOING:
+                            // wait until status change
+                            return Single.error(new IllegalStateException("ONGOING task " + taskId + " : trigger a retry"));
+                        default:
+                            // SUCCESS case
+                            return Single.just(true);
+                    }
+                }).retryWhen(new RetryWithDelay(3, 5000)).blockingGet();
+
+        if (!upgraded) {
+            throw getIllegalStateException();
+        }
+
+        return true;
+    }
+
+    protected abstract Single<Boolean> processUpgrade(String instanceOperationId, SystemTask task, String previousOperationId);
+
+    protected abstract IllegalStateException getIllegalStateException();
+
+    protected Single<SystemTask> createSystemTask(String operationId) {
+        final String taskId = getTaskId();
+        SystemTask systemTask = new SystemTask();
+        systemTask.setId(taskId);
+        systemTask.setType(SystemTaskTypes.UPGRADER.name());
+        systemTask.setStatus(SystemTaskStatus.INITIALIZED.name());
+        systemTask.setCreatedAt(new Date());
+        systemTask.setUpdatedAt(systemTask.getCreatedAt());
+        systemTask.setOperationId(operationId);
+        return systemTaskRepository.create(systemTask).onErrorResumeNext(err -> {
+            logger.warn("SystemTask {} can't be created due to '{}'", taskId, err.getMessage());
+            // if the creation fails, try to find the task, this will allow to manage the retry properly
+            return systemTaskRepository.findById(systemTask.getId()).toSingle();
+        });
+    }
+
+
+    protected abstract String getTaskId();
+
+    private class RetryWithDelay implements Function<Flowable<Throwable>, Publisher<?>> {
```

Comment:
This could be static, although I struggle to understand why this is flagged by intellij as a 'warning'.